working/0047. Spread Collections/spread-collections.md

```diff
@@ -0,0 +1,574 @@
+collection.
+
+The latter has some convenience appeal, but clashes with the rest of the
+language where null is never silently ignored. A null if statemt condition
+expression causes an exception instead of being implicitly treated as false as
+in most other languages.
+
+Most of the time, if you have a null in a place you don't expect, the sooner you
+can find out, the better. Even JavaScript does not silently ignore null in
+spreads. So I don't think we either. But, when looking through a corpus for
+places where a spread argument would be useful, I found a number of examples
+like:
+
+```dart
+var command = [
+  engineDartPath,
+  '--target=flutter',
+];
+if (extraFrontEndOptions != null) {
+  command.addAll(extraFrontEndOptions);
+}
+command.add(mainPath);
+```
+
+To handle these gracefully, we support a `...?` "null-aware spread" operator. In
+cases where the spread expression evaluates to null, that expands to an empty
+collection instead of throwing a runtime expression.
+
+That turns the example to:
+
+```dart
+var command = [
+  engineDartPath,
+  '--target=flutter',
+  ...?extraFrontEndOptions,
+  mainPath
+];
+```
+
+More complex conditional expressions than simple null checks come up often too,
+but those are out of scope for this proposal.
+
+## Syntax
+
+We extend the list grammar to allow *spread elements* in addition to regular
+elements:
+
+```
+listLiteral:
+  const? typeArguments? '[' listElementList? ']'
+  ;
+
+listElementList:
+  listElement ( ',' listElement )* ','?
+  ;
+
+listElement:
+  expression |
+  ( '...' | '...?' ) expression
+  ;
+```
+
+Instead of `expressionList`, this uses a new `listElementList` rule since
+`expressionList` is used elsewhere in the grammar where spreads aren't allowed.
+Each element in a list is either a normal expression or a *spread element*. If
+the spread element starts with `...?`, it's a *null-aware spread element*.
+
+The changes for map literals are similar:
+
+```
+mapLiteral:
+  const? typeArguments? '{' mapLiteralEntryList? '}' ;
+
+mapLiteralEntryList:
+  mapLiteralEntry ( ',' mapLiteralEntry )* ','?
+  ;
+
+mapLiteralEntry:
+  expression ':' expression |
+  ( '...' | '...?' ) expression
+  ;
+```
+
+Note that a *spread entry* for a map is an expression, not a key/value pair.
+Similar to lists, a spread entry that starts with `...?` is a *null-aware spread
+entry*.
+
+## Static Semantics
+
+Since the spread is unpacked and its individual elements added to the containing
+collection, we don't require the spread expression *itself* to be assignable to
+the collection's type. For example, this is allowed:
+
+```dart
+var numbers = <num>[1, 2, 3];
+var ints = <int>[...numbers];
+```
+
+This works because the individual elements in `numbers` do happen to have the
+right type even though the list that contains them does not. However, the spread
+object does need to be "spreadable"&mdash;it must be some kind of Iterable for a
+list literal or a Map for a map literal.
```

Comment:
Consider a comment to point out that this also doesn't require that the type of the thing being spread be assignable to the type of the List.  e.g. spreading MyList<Object> into a List<int> (neither is a subtype of the other).

Reply:
Done.